src/lib.rs

```diff
@@ -1378,6 +1382,289 @@ impl Default for Versioning {
+    /// let constraint_gt = VersioningReq::from_str(">=1.0.0").expect("Invalid req");
+    /// assert!(constraint_gt.matches(&Versioning::new("1.0.0").expect("Invalid version")));
+    /// assert!(constraint_gt.matches(&Versioning::new("1.1.0").expect("Invalid version")));
+    /// assert!(!constraint_gt.matches(&Versioning::new("0.9.0").expect("Invalid version")));
+    ///
+    /// let constraint_wild = VersioningReq::from_str("*").expect("Invalid req");
+    /// assert!(constraint_wild.matches(&Versioning::new("1.0.0").expect("Invalid version")));
+    /// assert!(constraint_wild.matches(&Versioning::new("1.1.0").expect("Invalid version")));
+    /// assert!(constraint_wild.matches(&Versioning::new("0.9.0").expect("Invalid version")));
+    ///
+    /// let constraint_eq = VersioningReq::from_str("==1.0.0").expect("Invalid req");
+    /// assert!(constraint_eq.matches(&Versioning::new("1.0.0").expect("Invalid version")));
+    /// assert!(!constraint_eq.matches(&Versioning::new("1.1.0").expect("Invalid version")));
+    /// assert!(!constraint_eq.matches(&Versioning::new("0.9.0").expect("Invalid version")));
+    /// ```
+    pub fn matches(&self, v: &Versioning) -> bool {
+        if let Some(sv) = &self.version {
+            match self.op {
+                Op::Exact => v == sv,
+                Op::Greater => v > sv,
+                Op::GreaterEq => v >= sv,
+                Op::Less => v < sv,
+                Op::LessEq => v <= sv,
+                Op::Tilde => self.matches_tilde(v),
+                Op::Caret => self.matches_caret(v),
+                Op::Wildcard => true,
+                _ => false,
+            }
+        } else {
+            matches!(self.op, Op::Wildcard)
+        }
+    }
+}
+
+impl FromStr for VersioningReq {
+    type Err = Error;
+
+    // NOTE: The parsing from `semver` is not `pub`, so we lift the pieces we need.
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let (op, s) = op(s)?;
+
+        let s = if s.is_empty() {
+            None
+        } else {
+            Versioning::new(s)
+        };
+
+        Ok(VersioningReq::new(op, s))
+    }
+}
+
+impl Default for VersioningReq {
+    fn default() -> Self {
+        Self::new(Op::Wildcard, None)
+    }
+}
+
+impl std::fmt::Display for VersioningReq {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            "{}{}",
+            // NOTE: Display is not implemented for `Op`
+            match self.op {
+                Op::Exact => "=",
+                Op::Greater => ">",
+                Op::GreaterEq => ">=",
+                Op::Less => "<",
+                Op::LessEq => "<=",
+                Op::Tilde => "~",
+                Op::Caret => "^",
+                Op::Wildcard => "",
+                _ => Err(std::fmt::Error)?,
+            },
+            self.version
+                .as_ref()
+                .map(|v| v.to_string())
+                .unwrap_or_default()
```

Comment:
These can be factored out as variables for that the `write!` occurs on a single line.